jax/core.py

```diff
@@ -493,9 +493,18 @@ def __array__(self, *args, **kw):
            "array indexing, like `x[idx]`, it may be that the array being "
            "indexed `x` is a raw numpy.ndarray while the indices `idx` are a "
            "JAX Tracer instance; in that case, you can instead write "
-           "`jax.device_put(x)[idx]`.")
+           "`jnp.asarray(x)[idx]`.")
     raise Exception(msg)
 
+  def __index__(self):
+    msg = (f"The __index__ method was called on the JAX Tracer object {self}.\n\n"
+           "This error can occur when a JAX Tracer object is used as an index for "
+           "a non-JIT-compatible object, such as a Python list or tuple. "
```

Comment:
I'm not sure if we should imply that lists and tuples are not jit-compatible, since they do work in many ways, just not with indexing.

How about something like:
"This error can occur when a JAX Tracer object is used as an index for a list or tuple."

Reply:
Sounds good. Turns out this also comes up when a tracer is used in `range()`, so I have to make the message a bit more contextual anyway!

Reply:
Oh great idea! That'll make those errors better too!